File: Algorithms/make-change.js
// Write a function that given an amount of money and a list of coin denominations computes the number of ways to make amount of money with coins of the available denominations.

function makeChange(amount, coins){

	var waysOfDoingNCents = []
	for (var i = 0; i < amount + 1; i++){
		waysOfDoingNCents[i] = 0;
	}

	waysOfDoingNCents[0] = 1;

	for (var coin = 0; coin < coins.length; coin++){
		for (var higherAmt = coins[coin]; higherAmt < amount + 1; higherAmt++){
			higherAmtRemainder = higherAmt - coins[coin];
			waysOfDoingNCents[higherAmt] += waysOfDoingNCents[higherAmtRemainder] 
		}
	}

	return waysOfDoingNCents[amount];
}

File: Algorithms/Cracking-the-Coding-Interview/make-change.rb
def make_change(amount, coins)
 result = Array.new(amount + 1, 0)
 result[0] = 1

 coins.each do |coin|
  (coin..amount).each do |higher_amount|
    remainder = higher_amount - coin
    result[higher_amount] += result[remainder]
  end
 end

 result[amount]
end